src/backend/env/mod.rs

```diff
@@ -1727,6 +1725,59 @@ impl State {
         });
     }
 
+    // Rewards a random user with a fixed amount of minted tokens.
+    // Users have a winning chance proportional to their weekly credits spending.
+    async fn random_reward() {
+        if let Ok((randomness,)) = raw_rand().await {
+            use std::convert::TryInto;
+            let bytes: [u8; 8] = randomness[0..8]
+                .try_into()
+                .expect("couldn't convert bytes to array");
+            let mut random_number: u64 = u64::from_be_bytes(bytes);
+
+            mutate(|state| {
+                // Creating random distribution of users with segments proportional to credits
+                // spent within the last week. Segments are placed randomly.
+                let mut allocation = Vec::new();
+                let mut threshold = 0;
+                for user in state.users.values_mut().filter(|user| {
+                    !user.controversial()
+                        && user.active_within_weeks(time(), 1)
+                        && user.credits_burned() > 0
+                }) {
+                    threshold += user.take_credits_burned();
+                    allocation.push((user.id, threshold));
+                }
+
+                // Truncate the random number so that every single user has a chance to win.
+                random_number %= threshold + 1;
```

Comment:
There seems to be some off-by-one issue in this distribution.

Let's take the simplest possible case of two users where each user burned 1 credit.

- threshold = 2
- random_number can take values 0, 1, 2
- the first user wins if the random number is 0 or 1
- the second user winds if the random number is 2

So the distribution is biased towards the first user.

A fix would be:
```
if threshold == 0 { bail out}
random_number %= threshold;
```

In the `find` predicate use the strict comparison: `threshold > random_number`.

Reply:
Yeah, sounds good to me! Thanks Ulan!